src/Workspaces/MSBuild/Core/MSBuild/BuildHostProcessManager.cs

```diff
@@ -221,16 +218,33 @@ private ProcessStartInfo CreateMonoBuildHostStartInfo(string pipeName)
 
     private static string GetDotNetFrameworkBuildHostPath()
     {
-        // The .NET Framework build host is deployed as a content folder next to the application into the BuildHost-net472 path
-        var netFrameworkBuildHost = Path.Combine(Path.GetDirectoryName(typeof(BuildHostProcessManager).Assembly.Location)!, "BuildHost-net472", "Microsoft.CodeAnalysis.Workspaces.MSBuild.BuildHost.exe");
-        AssertBuildHostExists(netFrameworkBuildHost);
-        return netFrameworkBuildHost;
+        return GetBuildHostPath("BuildHost-net472", "Microsoft.CodeAnalysis.Workspaces.MSBuild.BuildHost.exe");
     }
 
-    private static void AssertBuildHostExists(string buildHostPath)
+    private static string GetBuildHostPath(string contentFolderName, string assemblyName)
     {
-        if (!File.Exists(buildHostPath))
-            throw new Exception(string.Format(WorkspaceMSBuildResources.The_build_host_could_not_be_found_at_0, buildHostPath));
+        // Possible BuildHost paths are relative to where the Workspaces.MSBuild assembly was loaded.
+        var msbuildWorkspaceDirectory = Path.GetDirectoryName(typeof(BuildHostProcessManager).Assembly.Location)!;
+
+        // When Workspaces.MSBuild is deployed as part of an application the build host is deployed as a content folder next to the application
+        var buildHostPath = Path.Combine(msbuildWorkspaceDirectory, contentFolderName, assemblyName);
+        if (File.Exists(buildHostPath))
+        {
+            return buildHostPath;
+        }
+
+        // When Workspaces.MSBuild is loaded from the NuGet cache the Build host is deployed under the contentFiles folder inside the package
+        // 
+        // Workspaces.MSBuild.dll Path - .nuget/packages/microsoft.codeanalysis.workspaces.msbuild/{version}/lib/{tfm}/Microsoft.CodeAnalysis.Workspaces.MSBuild.dll
+        // MSBuild.BuildHost.dll Path  - .nuget/packages/microsoft.codeanalysis.workspaces.msbuild/{version}/contentFiles/any/any/{contentFolderName}/{assemblyName}
```

Comment:
I don't suppose there is any way to make this deterministic?  Like to know if the current msbuildworkspace dll is running from the nuget directory?

Reply:
NuGet folders are user configurable and there is also the possibility the nupkg was unzipped somewhere randomly. I think Jason suggested searching for a .nuspec to ensure we are in the scenario properly.